decidim-elections/app/commands/decidim/votings/census/admin/create_dataset.rb

```diff
@@ -64,27 +65,35 @@ def create_census_dataset!
           end
 
           def csv_header_invalid?
-            CSV.parse_line(File.open(file_path), col_sep: ";", headers: true, header_converters: :symbol).headers != expected_headers
+            headers.blank? || headers != expected_headers
           end
 
           def headers
+            @headers ||= CSV.parse_line(File.open(file_path), col_sep: ";", headers: true, header_converters: :symbol)&.headers
+          end
+
+          def no_ballot_headers
             [:document_id, :document_type, :date_of_birth, :full_name, :full_address, :postal_code, :mobile_phone_number, :email_address]
           end
 
           def ballot_style_headers
-            headers.push(:ballot_style_code)
+            no_ballot_headers.push(:ballot_style_code)
           end
 
           def expected_headers
-            @expected_headers ||= form.current_participatory_space.has_ballot_styles? ? ballot_style_headers : headers
+            @expected_headers ||= form.current_participatory_space.has_ballot_styles? ? ballot_style_headers : no_ballot_headers
           end
 
           def csv_row_count
             @csv_row_count ||= file_lines_count - 1
           end
 
+          # count lines in the most resource-efficient way using ruby, handles milions of lines with minimal memory footprint
           def file_lines_count
-            `wc -l "#{file_path.shellescape}"`.strip.split.first.to_i
+            # rubocop:disable Lint/EmptyBlock
+            File.foreach(file_path) {}
+            # rubocop:enable Lint/EmptyBlock
+            $INPUT_LINE_NUMBER
```

Comment:
I've made performance tests based on this article:
https://dalibornasevic.com/posts/68-processing-large-csv-files-with-ruby

My script was:

```ruby
require_relative './helpers'
require 'csv'

filename='data.csv'

print_memory_usage do
  print_time_spent do
    puts "\nwc"
    puts `wc -l "#{filename}"`.strip.split.first.to_i
  end
end

print_memory_usage do
  print_time_spent do
    puts "\nFile.open.readlines:"
    puts File.open(filename).readlines.size
  end
end

print_memory_usage do
  print_time_spent do
    puts "\nIO.readlines:"
    puts IO.readlines(filename).length
  end
end

print_memory_usage do
  print_time_spent do
    puts "\nFile.read/scan:"
    puts File.read(filename).scan(/\n/).length
  end
end

print_memory_usage do
  print_time_spent do
    puts "\nFile.foreach $.:"
    File.foreach(filename) {}
    puts $.
  end
end

print_memory_usage do
  print_time_spent do
    puts "\nIO.foreach $.:"
    IO.foreach(filename) { }
    puts $.
  end
end
``` 

Tested with 1 million lines, results are:

```
wc
1000001
Time: 0.02
Memory: 0.0 MB

File.open.readlines:
1000001
Time: 0.26
Memory: 141.84 MB

IO.readlines:
1000001
Time: 0.25
Memory: 7.8 MB

File.read/scan:
1000001
Time: 0.32
Memory: 74.5 MB

File.foreach $.:
1000001
Time: 0.22
Memory: 0.0 MB

IO.foreach $.:
1000001
Time: 0.21
Memory: 0.0 MB
```

Although the method chosen is abou 10 times slower than using `wc` it does not have a noticable memory impact and its portable.

Reply:
Really nice metrics, thanks a lot!

I'm not sure how you are measuring the memory usage, but I **think** there may be a small mistake with the `wc` performance metrics as I **think** you are only measuring the Ruby process' memory usage, not the separate process it runs for `wc`.

To get more accurate metrics for bash commands, use `/usr/bin/time` in bash:
```bash
$ /usr/bin/time -v wc -l million-lines.csv 
1000000 million-lines.csv
	Command being timed: "wc -l million-lines.csv"
	User time (seconds): 0.00
	System time (seconds): 0.00
	Percent of CPU this job got: 100%
	Elapsed (wall clock) time (h:mm:ss or m:ss): 0:00.01
	Average shared text size (kbytes): 0
	Average unshared data size (kbytes): 0
	Average stack size (kbytes): 0
	Average total size (kbytes): 0
	Maximum resident set size (kbytes): 2088
	Average resident set size (kbytes): 0
	Major (requiring I/O) page faults: 0
	Minor (reclaiming a frame) page faults: 109
	Voluntary context switches: 1
	Involuntary context switches: 3
	Swaps: 0
	File system inputs: 0
	File system outputs: 0
	Socket messages sent: 0
	Socket messages received: 0
	Signals delivered: 0
	Page size (bytes): 4096
	Exit status: 0
```

So you can see it actually consumes some amount of memory, although the average consumption is at zero as in your measurements.

The Ruby measurements seem correct. Not sure if the peak memory use is exactly at 0 but anyways as you already show, it's extremely low anyways so it shouldn't really matter.

Just one question regarding the actual code, why you did not just store the file number in a Ruby variable during processing? I don't see any significant increase in memory usage or time doing so. Or maybe my measurements don't work...

Reply:
> Well, the original code is not mine and it didn't occurred to me to refactor this. As it is now, the ActionLog and Dataset entries ares created before processing the CSV, that's why it needs to be counted separately. I think they wanted to do it this way to have registered all the attempts to upload a census, even if failed. I don't think it's a bad idea.

I think we might have misunderstood each other regarding that last point.

What I actually meant just to use a local variable to store the amount of lines instead of the global `$INPUT_LINE_NUMBER` variable as in:
```ruby
lines = 0
File.foreach(filename) { lines += 1 }
```

As the Rubocop violation is telling, using global variables is an anti-pattern and should be avoided everywhere we can. Most of the times they are only used in some of the command line scripts or in specs to refer to `$stdin` because we are not using any shell command library right now.

I tested this locally yesterday and I didn't see it having any significant performance impact regarding CPU time or memory usage.

Reply:
Ah, ok, right. If you prefer, but this is a builtin globl variable (that does this count for you). But yeah, it also might be more readable the other way (my original code used `$.` which is even more obscure).